Tribler/Test/Community/gigachannel/test_community.py

```diff
@@ -83,9 +85,31 @@ def test_send_random_multiple_torrents(self):
         yield self.deliver_messages(timeout=0.5)
 
         with db_session:
-            self.assertEqual(len(self.nodes[1].overlay.metadata_store.ChannelMetadata.select()), 1)
-            channel = self.nodes[1].overlay.metadata_store.ChannelMetadata.select()[:][0]
+            channel = self.nodes[1].overlay.metadata_store.ChannelMetadata.get()
+            torrents1 = self.nodes[1].overlay.metadata_store.TorrentMetadata.select()[:]
             self.assertLess(channel.contents_len, 20)
+            self.assertLess(0, channel.contents_len)
+
+        self.assertEqual(1, self.nodes[0].overlay.gossip_sequence_count)
+
+        # Initiate the gossip again. This time, it should be sent from the blob cache
+        # so the torrents on the receiving end should not change this time.
+        self.nodes[0].overlay.send_random_to(Peer(self.nodes[1].my_peer.public_key, self.nodes[1].endpoint.wan_address))
+
+        yield self.deliver_messages(timeout=0.5)
+        with db_session:
+            torrents2 = self.nodes[1].overlay.metadata_store.TorrentMetadata.select()[:]
+            self.assertEqual(len(torrents1), len(torrents2))
+
+        self.nodes[0].overlay.send_random_to(Peer(self.nodes[1].my_peer.public_key, self.nodes[1].endpoint.wan_address))
+
+        # Warning! The following test can fail once in a blue moon because the second batch of gossiped torrents
```

Comment:
Please, dear god, no racy tests. It's better to have no test, than a test which randomly fails.

Reply:
The blue moon is not going to happen soon. The probability of the test failing with the set of 40 items of which we randomly select 4 is: ` 4! / (40!/36!)`, which is 0,000010942.
If that is not good enough for you, I can make the probability of the fail as high as you like :wink:

Reply:
Windows probably ignores this blue moon, it's really better to have uncovered code rather than having tests that might fail, even if the probability appears to be small.

Reply:
> Windows probably ignores this blue moon, it's really better to have uncovered code rather than having tests that might fail, even if the probability appears to be small.

I do not agree. If a test that should fail with some marginally low probability fails more than once in our lifetime, that is a good indication of something fishy going on that should be investigated.

Reply:
https://hackernoon.com/flaky-tests-a-war-that-never-ends-9aa32fdef359
https://martinfowler.com/articles/nonDeterminism.html

Reply:
@qstokkink , I hate non-determinism as much as you do. However, as a trained physicist, I humbly understand that every thing in life is only determined to a point :wink: 

From the hackernoon article:
> "Usually with these tests a non-deterministic failure is relatively common, so you end up shrugging your shoulders when these tests go red."

A year has 365 days. Let's say we run 100 tests a day. Let's say we continue development for 20 years. If I add 200 torrents, we go to `0,000000015*100*36500*20 = 1` expected fail during these 20 years. For me, that is a fair price for having a working integration test (that can become deterministic at some point, when we learn how to control RNG seeds in Python tests...).

So if you think that the price is too high - I'll remove the test.